pageserver/src/tenant/vectored_blob_io.rs

```diff
@@ -337,38 +339,68 @@ impl<'a> VectoredBlobReader<'a> {
                 .chain(std::iter::once(None)),
         );
 
+        // Some scratch space, put here for reusing the allocation
+        let mut decompressed_vec = Vec::new();
+
         for ((offset, meta), next) in pairs {
             let offset_in_buf = offset - start_offset;
             let first_len_byte = buf[offset_in_buf as usize];
 
-            // Each blob is prefixed by a header containing it's size.
+            // Each blob is prefixed by a header containing its size and compression information.
             // Extract the size and skip that header to find the start of the data.
             // The size can be 1 or 4 bytes. The most significant bit is 0 in the
             // 1 byte case and 1 in the 4 byte case.
-            let (size_length, blob_size) = if first_len_byte < 0x80 {
-                (1, first_len_byte as u64)
+            let (size_length, blob_size, compression_bits) = if first_len_byte < 0x80 {
+                (1, first_len_byte as u64, BYTE_UNCOMPRESSED)
             } else {
                 let mut blob_size_buf = [0u8; 4];
                 let offset_in_buf = offset_in_buf as usize;
 
                 blob_size_buf.copy_from_slice(&buf[offset_in_buf..offset_in_buf + 4]);
-                blob_size_buf[0] &= 0x7f;
-                (4, u32::from_be_bytes(blob_size_buf) as u64)
+                blob_size_buf[0] &= !LEN_COMPRESSION_BIT_MASK;
+
+                let compression_bits = first_len_byte & LEN_COMPRESSION_BIT_MASK;
+                (
+                    4,
+                    u32::from_be_bytes(blob_size_buf) as u64,
+                    compression_bits,
+                )
             };
 
-            let start = offset_in_buf + size_length;
-            let end = match next {
+            let start_raw = offset_in_buf + size_length;
+            let end_raw = match next {
                 Some((next_blob_start_offset, _)) => next_blob_start_offset - start_offset,
-                None => start + blob_size,
+                None => start_raw + blob_size,
             };
-
-            assert_eq!(end - start, blob_size);
+            assert_eq!(end_raw - start_raw, blob_size);
+            let (start, end);
+            if compression_bits == BYTE_UNCOMPRESSED {
```

Comment:
The two pieces of code are equivalent, but it's due to different way of how `compression_bits` is written to.

In `blob_io`, we do `compression_bits = first_len_byte & LEN_COMPRESSION_BIT_MASK`, so `compression_bits` goes from 0x10 to 0xf0 potentially. Anything `< BYTE_UNCOMPRESSED`, so anything where that single bit is not set, is still valid input. It just came from the case where there is a single byte for the length.

On the other hand, in `vectored_blob_io`, we do `compression_bits = BYTE_UNCOMPRESSED` if there is only a single byte for the length and `compression_bits = first_len_byte & LEN_COMPRESSION_BIT_MASK` otherwise.

I actually followed Joonas' advice first to change that to `==` instead, but it failed the CI tests and then I reverted it to the original state.

It's a bit hard to notice I admit, but there is subtle differences in between `vectored_blob_io` and `blob_io` modules, so I think it's fine. Shrug, in the medium to long term `blob_io`'s reader will be gone anyway, no?